### PR TITLE
Add validation to ensure valid moduleCode

### LINKED ISSUE #70: [PE-D][Tester B] Duplicate modules accepted
added cs2040s and then added CS2040S to index 1 and it seems to accept duplicate modules. It might be better to see this as a feature flaw so users don't confuse themselves in some cases.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: Hwww23/ped#1

### LINKED ISSUE #72: [PE-D][Tester B] Unable to remove module that have capitalization
Tried removing cs2040S from index 1 with modules(cs2040s and CS2040S)

It could help users if it doesn't check for differences in capitalization when removing a module since the letters and numbers for modules are unique.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: Hwww23/ped#2

### LINKED ISSUE #78: [PE-D][Tester D] List_modules Command not working
As shown in the screenshot, using list_modules m/cs2030 does not work as intended. There could be an issue with the list_modules command.

-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: justinlengch/ped#2

### LINKED ISSUE #84: [PE-D][Tester B] Unable to find student taking the specified module when not using capitalization
It does not list out students taking CS2040S/cs2040s, if I use Cs2040s. This could be a feature flaw since module code names have a unique combination of letters and numbers which should not have an issue with/without capitalization. This could benefit users from having to type the module name very specifically.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: Hwww23/ped#7

### LINKED ISSUE #91: [PE-D][Tester A] list_modules doesn't seem to be working.
Command: list_modules m/CS

The command list_modules doesn't seem to be working. I tried to find any modules with CS, but it doesn't do anything.

Command: list_modules m/CS2103S

When I tried to find the specific module CS2103S and adding a timing for 1 student, it still doesn't work.

Command: list_modules m/CS3323

I even tried it with a module I haven't added and it doesn't work.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: DominicFuMJ/ped#6

### LINKED ISSUE #104: [PE-D][Tester D] Module code for add_timing command is case-sensitive
Even though it is the same module code, add_timing does not work, saying module not found, even when the module exist for the student. The user guide should mention that it is case-sensitive.

-------------
Labels: `severity.Low` `type.DocumentationBug`
original: justinlengch/ped#8

### LINKED ISSUE #105: [PE-D][Tester C] MODULE_CODE Input Accepts Invalid Formats Contradicting NUS Module Code Formats
The field for entering a MODULE_CODE does not enforce the expected format based on NUS module codes.  It allows inputs that are clearly invalid (e.g., "QWER1234SSSS"). Such invalid module codes should be rejected with an informative error message

-------------
Labels: `severity.Low` `type.FeatureFlaw`
original: ejnan/ped#10

### LINKED ISSUE #108: [PE-D][Tester B] module_search accepts invalid module
Since module_search m/cs203s check for module format, user should expect longer invalid format to not work as well. However, module_search m/cs20300s runs successfully. This might be a feature flaw since module names are usually coded from 1000 to 8000 with optional letters behind.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: Hwww23/ped#10

### LINKED ISSUE #110: [PE-D][Tester B] Accepts adding module
Using add_module i/1 m/cs20300s runs successfully. This could be a feature flaw since module names are coded from 1000 to 8000.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: Hwww23/ped#11

### LINKED ISSUE #122: [PE-D][Tester D] List_modules command accepts empty string as module code
Shown above, an empty string can be searched, but does not make sense when considering add_module command does not accept an empty string, hence there would not be a module with an empty string as its name in the first place.

-------------
Labels: `type.FeatureFlaw` `severity.Medium`
original: justinlengch/ped#19

### PR DESCRIPTION
**Pull request by yadunut**
This PR closes #70. 
By proxy, this PR also closes #72, closes #84. However would like to double check for this one. 
Closes #110.
Closes #108.
Closes #91.
Closes #105.
Closes #122.
Closes #104.
Closes #78.

### CODE REVIEW

**Comment by yadunut** [comment]
So this hinges on the fact that I can get reading module_data working in shadowJar files...

**Comment by yadunut** [comment]
Fixed the shadowjar issue. Whoever is reviewing this, if you don't mind, pls checkout this branch and run

1. `./gradlew shadowJar`
2. `cd build/libs`
3. java -jar addressbook.jar
4. `list_modules m/cs` should return a list of modules

File: src/main/java/seedu/address/logic/parser/ListModulesCommandParser.java [hunk 1]
```diff
@@ -22,6 +22,10 @@ public ListModulesCommand parse(String userInput) throws ParseException {
         }
         argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_MODULE_CODE);
         String modulePrefix = argMultimap.getValue(PREFIX_MODULE_CODE).get();
+        if (modulePrefix.trim().isEmpty()) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, ListModulesCommand.MESSAGE_USAGE));
+        }
```
Comment (blaukc) [inline src/main/java/seedu/address/logic/parser/ListModulesCommandParser.java]: should this be in `ParserUtil.java`? where a `parseModulePrefix` returns the trimmed string? but it seems to work even if i have whitespace ard the moduleCode though so it is still functional

**Review by blaukc — APPROVED** [review]
image

your list_modules works, lgtm

File: src/main/java/seedu/address/logic/parser/ListModulesCommandParser.java (hunk 1 above)
Comment (yadunut) [reply src/main/java/seedu/address/logic/parser/ListModulesCommandParser.java]: Nice catch!

**Comment by AdityaB4** [comment]
Note to graders: while this is a validity check update, it is allowed according to the guidelines since it causes the search features to misbehave. Refer to Q5 under FAQs for more details.

**Pull request merged**